src/util/git_repos.rs

```diff
@@ -1,54 +1,288 @@
+            .arg(format!(
+                "git rev-parse --verify \"{}^{{commit}}\"",
+                commit_maybe
+            ))
+            .output()?;
+
+        // If the command was successful, return true
+        if output.status.success() {
+            Ok(true)
+        } else {
+            let stderr = String::from_utf8_lossy(&*output.stderr).to_string();
+            // If stderr matches the message that git spits out if the commit is not valid, return
+            // false
+            if stderr.trim() == "fatal: Needed a single revision" {
+                return Ok(false);
+            }
+            // Otherwise return an error
+            Err(Error::Git(stderr))
+        }
+    }
 
+    /// Uses git show-ref to verify whether `tag_maybe` is a tag.  Returns true if it is, or false
+    /// if it's not.  Returns an error if the command fails for some reason.
+    fn git_show_ref_verify(&self, repo_dir: &Path, tag_maybe: &str) -> Result<bool, Error> {
+        // Run the command
         let output = Command::new("sh")
+            .current_dir(repo_dir)
             .arg("-c")
-            .arg(format!("git ls-remote {}", url_to_check))
+            .arg(format!("git show-ref --verify \"refs/tags/{}\"", tag_maybe))
             .output()?;
 
+        // If the command was successful, return true
         if output.status.success() {
             Ok(true)
         } else {
-            Ok(false)
+            let stderr = String::from_utf8_lossy(&*output.stderr).to_string();
+            // If stderr says it's not a valid ref, we'll just return false
+            if stderr.contains("- not a valid ref") {
+                return Ok(false);
+            }
+            // Otherwise return an error
+            Err(Error::Git(stderr))
+        }
+    }
+
+    /// Calls git rev-list on the repo in `repo_dir` for `tag` and returns the resulting commit if
+    /// successful, or an error if the command fails for some other reason
+    fn git_rev_list(&self, repo_dir: &Path, tag: &str) -> Result<String, Error> {
+        // Run the command
+        let output = Command::new("sh")
+            .current_dir(repo_dir)
+            .arg("-c")
+            .arg(format!("git rev-list -n 1 {}", tag))
+            .output()?;
+
+        // If the command was successful, return the commit string
+        if output.status.success() {
+            Ok(String::from_utf8_lossy(&*output.stdout)
+                .to_string()
+                .trim()
+                .to_string())
+        } else {
+            // Otherwise return an error
+            Err(Error::Git(
+                String::from_utf8_lossy(&*output.stderr).to_string(),
+            ))
+        }
+    }
+
+    /// Calls git tag --points-at on the repo in `repo_dir` for `commit` and returns the resulting
+    /// tags if successful, or an error if the command fails for some other reason
+    fn git_tag_points_at(&self, repo_dir: &Path, commit: &str) -> Result<Vec<String>, Error> {
```

Comment:
Did you consider using something like the `git2` crate to do this? (https://docs.rs/git2/latest/git2/)

Reply:
I did and I ended up deciding not to use it because it doesn't seem to support some of the options I'm using in some of the commands.  I might take another look at it later on, though, because it would definitely be preferable to running sh every time.

Reply:
👍 

Sounds good.  I figured there was a reason, but I couldn't remember what it was.